vendor/github.com/dgryski/go-tsz/tsz.go

```diff
@@ -276,6 +276,30 @@ func (it *Iter) Next() bool {
 
 		it.val = math.Float64frombits(v)
 
+		// special case: read upcoming dod
+		// if delta+dod <0, the delta overflowed, and should rectify it.
+		// see https://github.com/grafana/metrictank/pull/1126
+		// but we must take a backup of the stream because reading from the
+		// stream modifies it.
+		brBackup := it.br.clone()
+		dod, ok := it.dod()
+		if !ok {
+			// in this case we can't know if the point is right or wrong.
+			// long chunks with a single point in them may lead to a wrong read, but this should be rare.
+			// we can't just adjust the timestamp because we don't know the length of the chunk
+			// (though this could be done by having the caller pass us that information), nor whether
+			// a delta that may seem low compared to chunk length was intentional or not.
+			// so, nothing much to do in this case. return the possibly incorrect point.
+			// and for return value, stick to normal iter semantics:
+			// this read succeeded, though we already know the next one will fail
+			it.br = *brBackup
+			return true
+		}
+		if dod+int32(tDelta) < 0 {
+			it.tDelta += 16384
```

Comment:
correct.
note that the plan is to, after putting the remediation in place, deprecate this chunk format asap (at least for >4h chunks) and start using a chunk format that doesn't have this bug.

Reply:
I wonder if that copying of `.br` could be saved if `.dod()` would have a small read buffer that can store 1 bit. If there's a value in the read buffer when `.dod()` is called it returns the content of the read buffer and clears it, otherwise it goes to the bstream as it does now. After that first call site of `.dod()` where we know that on the next call to `.dod()` we'll need to get the same value returned, we could then just put that value into the read buffer. That would save the copying of that bstream, but it makes everything a little more complicated, so it might not be worth it.

Reply:
That read buffer could be just two `bool` values. One is the actual value that's been buffered and the other one indicates if the buffer is empty or has a value.